builder demotion for block validation errors (#283)

tosynthegeek opened the pull request:
This PR addresses issue #283 by implementing builder demotion for `BlockValidationFailed` errors (e.g., block hash mismatch). The changes ensure that the builder for the invalid block is demoted once per incident, using both cache (`local_cache.demote_builder`) and database (`db.db_demote_builder`) for immediate and persistent effects, respectively.

## Changes
- Added `builder_pubkey: Option<BlsPublicKeyBytes>` to `GetPayloadResultData` to pass the builder's public key from `bid_sorter.get_builder_pub_key`.
- Modified `handle_get_payload` to retrieve `builder_pubkey` from `ForkState` using `parent_hash`.
- Updated `ProposerApi::_get_payload` to demote builders on `BlockValidationFailed` errors after `MultiBeaconClient::publish_block`.

ltitanb commented on crates/relay/src/auctioneer/bid_sorter.rs:
```diff
@@ -224,6 +224,10 @@ impl BidSorter {
         self.forks.get(parent_hash).and_then(|s| s.curr_bid.as_ref().map(|b| b.1.block_hash))
     }
 
+    pub fn get_builder_pub_key(&self, parent_hash: &B256) -> Option<BlsPublicKeyBytes> {
```
the builder pubkey should be already in the "local" payload we have in the payload entry, if not let's add it there instead of getting it from the bid sorter

tosynthegeek replied:
> the builder pubkey should be already in the "local" payload we have in the payload entry, if not let's add it there instead of getting it from the bid sorter

Okay I will add `builder_pubkey` to `PayloadEntry` and pass it through to `handle_get_payload`, but since `new_gossip` payloads may not include a builder key, I may have to keep it as optional

tosynthegeek replied:
> > the builder pubkey should be already in the "local" payload we have in the payload entry, if not let's add it there instead of getting it from the bid sorter
> 
> Okay I will add `builder_pubkey` to `PayloadEntry` and pass it through to `handle_get_payload`, but since `new_gossip` payloads may not include a builder key, I may have to keep it as optional

@ltitanb please let me know

ltitanb commented on crates/relay/src/api/proposer/get_payload.rs:
```diff
@@ -283,6 +298,39 @@ impl<A: Api> ProposerApi<A> {
+                                        %err,
+                                        %block_hash_clone,
+                                        "failed to demote builder in database! Pausing all optimistic submissions"
+                                    );
```
here you are not really pausing submission right? this is a copy of the demotion logic in `handle_simulation_result` however id rather we keep the demotion in one place. Let's make a new event , "BuilderDemotion" to be sent to the auctioneer , and refactor the existing logic to be used by both that event and the `handle_simulation_result`

tosynthegeek replied:
> here you are not really pausing submission right? this is a copy of the demotion logic in `handle_simulation_result` however id rather we keep the demotion in one place. Let's make a new event , "BuilderDemotion" to be sent to the auctioneer , and refactor the existing logic to be used by both that event and the `handle_simulation_result`

Make sense, I will resume this once I get clarification from the first review

ltitanb commented:
hey @tosynthegeek , yes apologies for the delay! If you could please resolve the branch conflicts then me and @Owen66 will review!

tosynthegeek committed:
feat: add builder demotion trigger on block validation failure
- Introduced `BuilderDemotion` event to trigger when `BeaconClientError::BlockValidationFailed` occurs
- Added `send_event` method to `AuctioneerHandle` for safe event dispatch using private auctioneer sender
- Used `builder_pubkey` from payload bid data to identify the demoted builder
- Implemented `handle_builder_demotion` in `Context` to process demotion events
